### PR TITLE
Implement str.isascii() based on CPython

### PR DESCRIPTION
**Pull request by 1e-to**

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. Seems like CI is currently failing so will mark as "in progress".

**Comment by 1e-to** [comment]
> Thanks for the PR. Seems like CI is currently failing so will mark as "in progress".
Method str.isascii() appears in python version 3.7, I cant do green CI on machines with python v < 3.7
On other machines, the function works correctly

**Comment by stuartarchibald** [comment]
I think you'll need to add a feature guard for the implementation like:
```python
if sys.version_info[:2] >= (3, 7):
  @overload_method(types.UnicodeType, 'isascii')
  def ... 
```
and similar to skip testing on unsupported python version, example:
https://github.com/numba/numba/blob/3fba0fbbef840aa1d055130cf0574adfeade2c82/numba/tests/test_unicode.py#L313-L315

**Comment by 1e-to** [comment]
Add to queue for review, please

File: numba/unicode.py [hunk 1]
```diff
@@ -1388,6 +1389,20 @@ def impl(s):
...
 
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11729-L11737    # noqa: E501
+if sys.version_info[:2] >= (3, 7):
+    @overload_method(types.UnicodeType, 'isascii')
+    def unicode_isalpha(data):
+        """Implements UnicodeType.isascii()"""
+
+        def impl(data):
+            if data._is_ascii:
```
Comment (stuartarchibald) [inline numba/unicode.py]: how about just doing `return data._is_ascii` ?

File: numba/unicode.py [hunk 2]
```diff
@@ -1388,6 +1389,20 @@ def impl(s):
     return impl
 
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11729-L11737    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode.py]: I'm not sure how the source code referenced by this citation bears any relation to the implementation below?

File: numba/tests/test_unicode.py [hunk 3]
```diff
@@ -1277,6 +1278,23 @@ def pyfunc(x):
...
+            return x.isascii()
+
+        cfunc = njit(pyfunc)
+        # Samples taken from CPython testing:
+        # https://github.com/python/cpython/blob/865c3b257fe38154a4320c7ee6afb416f665b9c2/Lib/test/string_tests.py#L913-L926     # noqa: E501
+        cpython = ['', '\x00', '\x7f', '\x00\x7f', '\x80', '\xe9']
+
+        msg = 'Results of "{}".isascii() must be equal'
+        for s in UNICODE_EXAMPLES + [''] + cpython:
+            for i in range(8):
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: What is the purpose of this loop?

File: numba/unicode.py [hunk 4]
```diff
@@ -1388,6 +1389,20 @@ def impl(s):
...
+if sys.version_info[:2] >= (3, 7):
+    @overload_method(types.UnicodeType, 'isascii')
+    def unicode_isalpha(data):
```
Comment (stuartarchibald) [inline numba/unicode.py]:
```suggestion
    def unicode_isascii(data):
```
please be careful with naming.

File: numba/unicode.py (hunk 1 above)
Comment (1e-to) [reply numba/unicode.py]: Fixed

File: numba/unicode.py (hunk 2 above)
Comment (1e-to) [reply numba/unicode.py]: There is no relation, I thought it was necessary in any case to leave a link to the implementation in cpyton
Removed

File: numba/tests/test_unicode.py (hunk 3 above)
Comment (1e-to) [reply numba/tests/test_unicode.py]: In сpython there is optimization for reading 4 and 8 bytes at once, we do not have it
Removed

File: numba/unicode.py (hunk 4 above)
Comment (1e-to) [reply numba/unicode.py]: Fixed

File: numba/unicode.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/unicode.py]: The links are algorithmic references and related to licensing:
https://github.com/numba/numba/blob/e4674eea90b451ae2141a20546727bef4ac8b352/LICENSES.third-party#L173-L179

**Pull request merged**